src/ddmd/expressionsem.d

```diff
@@ -7924,59 +7924,61 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
         exp.setNoderefOperands();
 
         // same as for AndAnd
-        exp.e1 = exp.e1.semantic(sc);
-        exp.e1 = resolveProperties(sc, exp.e1);
-        exp.e1 = exp.e1.toBoolean(sc);
+        Expression e1x = exp.e1.semantic(sc);
+        e1x = resolveProperties(sc, e1x);
+        e1x = e1x.toBoolean(sc);
         uint cs1 = sc.callSuper;
 
         if (sc.flags & SCOPEcondition)
         {
             /* If in static if, don't evaluate e2 if we don't have to.
              */
-            exp.e1 = exp.e1.optimize(WANTvalue);
-            if (exp.e1.isBool(true))
+            e1x = e1x.optimize(WANTvalue);
+            if (e1x.isBool(true))
             {
                 result = new IntegerExp(exp.loc, 1, Type.tbool);
                 return;
             }
         }
 
-        exp.e2 = exp.e2.semantic(sc);
+        Expression e2x = exp.e2.semantic(sc);
         sc.mergeCallSuper(exp.loc, cs1);
-        exp.e2 = resolveProperties(sc, exp.e2);
+        e2x = resolveProperties(sc, e2x);
 
-        auto f1 = checkNonAssignmentArrayOp(exp.e1);
-        auto f2 = checkNonAssignmentArrayOp(exp.e2);
+        auto f1 = checkNonAssignmentArrayOp(e1x);
+        auto f2 = checkNonAssignmentArrayOp(e2x);
         if (f1 || f2)
         {
             result = new ErrorExp();
             return;
         }
 
-        if (exp.e2.type.ty == Tvoid)
+        if (e2x.type.ty == Tvoid)
```

Comment:
By assigning to `exp.type` here, the type gets disconnected from the operands, putting `exp` into an invalid state. A better approach is to assign to a `t` here, and assign `exp.type = t;` when the operands are assigned.